server/routerlicious/packages/services/src/mongoExceptionRetryRules/mongoError/index.ts

```diff
@@ -0,0 +1,172 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IMongoExceptionRetryRule } from "../IMongoExceptionRetryRule";
+class InternalErrorRule implements IMongoExceptionRetryRule {
+    private static readonly codeName = "InternalError";
+    match(error: any): boolean {
+        return error.code === 1
+            && error.codeName
+            && (error.codeName as string) === InternalErrorRule.codeName;
+    }
+
+    shouldRetry: boolean = false;
+}
+
+// This handles the requested queued on client side buffer overflow. Should relies on reconnect instead of retry?
+class NoConnectionAvailableRule implements IMongoExceptionRetryRule {
+    private static readonly messagePrefix = "no connection available for operation and number of stored operation";
+    match(error: any): boolean {
+        // TODO: This timed out actually included two different messages:
+        // 1. Retries due to rate limiting: False.
+        // 2. Retries due to rate limiting: True.
+        // We might need to split this into two different rules after consult with DB team.
+        return error.message
+            && (error.message as string).startsWith(NoConnectionAvailableRule.messagePrefix);
```

Comment:
Is it possible that the `error.message` starts with capital letter? Do we need to make this comparison case insensitive?

Reply:
I think for now we should be strict as much as possbile and log the unmatched rule for `DefaultRule`, and we can extend the ruleset further.